rust/sealevel/libraries/account-utils/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+/// Consider removing the `Default` binding in the future.
+pub trait Data: BorshDeserialize + BorshSerialize + Default {}
+
+impl<T> Data for T where T: BorshDeserialize + BorshSerialize + Default {}
+
+/// Account data structure wrapper type that handles initialization and (de)serialization.
+///
+/// (De)serialization is done with borsh and the "on-disk" format is as follows:
+/// {
+///     initialized: bool,
+///     data: T,
+/// }
+#[derive(Debug, Default)]
+pub struct AccountData<T> {
+    data: Box<T>,
+}
+
+impl<T> From<T> for AccountData<T> {
+    fn from(data: T) -> Self {
+        Self {
+            data: Box::new(data),
+        }
+    }
+}
+
+impl<T> From<Box<T>> for AccountData<T> {
+    fn from(data: Box<T>) -> Self {
+        Self { data }
+    }
+}
+
+impl<T> SizedData for AccountData<T>
+where
+    T: SizedData,
+{
+    fn size(&self) -> usize {
+        // Add an extra byte for the initialized flag.
+        1 + self.data.size()
+    }
+}
+
+impl<T> AccountData<T>
+where
+    T: Data,
+{
+    pub fn into_inner(self) -> Box<T> {
+        self.data
+    }
+
+    /// Deserializes the account data from the given slice.
+    pub fn fetch_data(buf: &mut &[u8]) -> Result<Option<Box<T>>, ProgramError> {
+        if buf.is_empty() {
+            return Ok(None);
+        }
+        // Account data is zero initialized.
+        let initialized = bool::deserialize(buf)?;
+        let data = if initialized {
+            Some(T::deserialize(buf).map(Box::new)?)
+        } else {
+            None
+        };
+        Ok(data)
+    }
+
+    /// Deserializes the account data from the given slice and wraps it in an `AccountData`.
+    pub fn fetch(buf: &mut &[u8]) -> Result<Self, ProgramError> {
+        Ok(Self::from(Self::fetch_data(buf)?.unwrap_or_default()))
+    }
+
+    // Optimisically write then realloc on failure.
+    // If we serialize and calculate len before realloc we will waste heap space as there is no
+    // free(). Tradeoff between heap usage and compute budget.
+    pub fn store(
+        &self,
+        account: &AccountInfo<'_>,
+        allow_realloc: bool,
+    ) -> Result<(), ProgramError> {
+        if !account.is_writable || account.executable {
+            return Err(ProgramError::InvalidAccountData);
+        }
+        let realloc_increment = 1024;
+        loop {
+            let mut guard = account.try_borrow_mut_data()?;
+            let data = &mut *guard;
+            let data_len = data.len();
+
+            match self.store_in_slice(data) {
+                Ok(_) => break,
+                Err(err) => match err.kind() {
+                    std::io::ErrorKind::WriteZero => {
+                        if !allow_realloc {
+                            return Err(ProgramError::BorshIoError(err.to_string()));
+                        }
+                    }
+                    _ => return Err(ProgramError::BorshIoError(err.to_string())),
+                },
+            };
+            drop(guard);
```

Comment:
I personally prefer dropping guards by exiting a block, because it helps to more easily visualise for how long the guard is held. If that makes sense to you, I would refactor this into a block that returns `data_len` so the `guard` is automaticelly dropped. Could also be a function